### ISSUE TITLE
Allow custom types to be mapped to Realm primitive types

### ISSUE THREAD

**Issue by jpsim** [realm/realm-swift#3792]
This could allow types like NSURL (#599), unsigned integers (#651) and UUIDs (#2256) among others.

One way to do this for both Objective-C and Swift would be to expose a protocol (e.g. `RealmConvertible`) that exposes a declarative interface to map the equivalent Realm primitives required to represent the type.

``` objc
@protocol RLMRealmConvertible
+ (NSArray<RLMProperty *> *)realmProperties;
- (RLMObject *)toRealmObject;
- (instancetype)initWithRealmObject:(RLMObject *)object;
@end
```

^this strawman interface probably makes no sense, but is just to kick off a discussion.

**Comment by JadenGeller** [comment]
If users were to modify a field on these types, how would Realm know to update the datastore representation of the object? It seems like we'd only be able to support "immutable" types.

**Comment by JadenGeller** [comment]
Though I guess we could KVO relevant properties if the object is KVO compliant.

**Comment by azinman** [comment]
+1 for KVO.

**Comment by tgoyne** [comment]
This was implemented in 10.21.0
